book/scheduling-and-threading.md

```diff
@@ -312,256 +327,241 @@ situation---some devices have more [CPU cores][cores] than others, or are more
-            style(self.nodes, sorted(self.rules,
-                key=cascade_priority))
-            self.document = DocumentLayout(self.nodes)
-            self.document.layout()
-            self.display_list = []
-            self.document.paint(self.display_list)
-            self.needs_pipeline_update = False
-
 class Browser:
     def raster_and_draw(self):
         self.raster_chrome()
         self.raster_tab()
         self.draw()
 ```
 
-The last piece is to actually call `run_animation_frame` from somewhere; at the
-moment all we're doing is setting `needs_animation frame`. The place to put
-it is in `run_once`, of course:
-
-``` {.python expected=False}
-class TaskRunner:
-    # ...
-    def run_once():
-        # ...
-        lock.acquire(blocking=True)
-        needs_animation_frame = self.needs_animation_frame
-        lock.release()
-        if needs_animation_frame:
-            self.tab.run_animation_frame()
-```
-
-Rendering is now async and scheduled according to the desired cadence.
-
-Let's now take advantage of this new asynchronous technology by replacing all
-cases where the rendering pipeline is computed synchronously with
-`set_needs_pipeline_update`. Here is `load`, for example:[^more-examples]
+The last piece is to actually call `set_needs_render` from somewhere.
+Replace all cases where the rendering pipeline is computed synchronously with
+`set_needs_render`. Here, for example, is `load`:[^more-examples]
 
 [^more-examples]: There are more of them; you should fix them all.
 
 ``` {.python}
 class Tab:
     def load(self, url, body=None):
         # ...
-        self.set_needs_pipeline_update()
+        self.set_needs_render()
 ```
 
-Now our browser can run rendering in an asynchronous, scheduled way on the event
-loop!
-
-But along the way we lost some performance. We're now always calling
-`raster_and_draw` instead of only when needed. A dirty bit called
-`needs_raster_and_draw` on `Browser` can fix most of this. Wherever it used to
-call `raster_chrome`, `raster_tab` or `draw`  directly, now the browser should
-set the dirty bit.^[Note that this will not get back some of the same
-performance we had at the end of
-[chapter 11](visual-effects.md#browser-compositing), where the browser only ran
-raster and draw when needed. I've left this as an exercise.]
-
-``` {.python} 
-class Browser:
-    def __init__(self):
-        # ...
-        self.needs_raster_and_draw = False
-
-    def set_needs_raster_and_draw(self):
-        self.needs_raster_and_draw = True
-
-    def raster_and_draw(self):
-        if not self.needs_raster_and_draw:
-            return
-        # ...
-        self.needs_raster_and_draw = False
-
-```
-
-Here's the change to `handle_click` to set the dirty bit instead of calling
-`raster_chrome` or `raster_tab` directly:
+All the places that call `raster_chrome`, `raster_tab` or `draw` directly will
+also need to call `set_needs_render` instead. Here's `handle_down`:
```

Comment:
Is this actually necessary? I guess so for "raster_tab" but not for "raster_chrome" or "draw", right? I'm not too sure.

Reply:
It's not "necessary", but I put it in because they should think of the entire pipeline as one thing. I added a footnote.